### PR TITLE
Add "grouped diagnostics" that allows rendering across source files 

### PR DESCRIPTION
**Pull request by DougGregor**
Extend the diagnostics renderer to handle a set of "grouped"
diagnostics, where the diagnostics occur across different source files
that are related through source code generation, for example because
one is a macro expansion buffer that was triggered within the other.
These inner source files are shown as nested source files in the
output, underneath the location of where the outer source file where
they logically occur.

Introduce a few other small improvements to the colorization in the 
diagnostics formatter that became more apparent when we do this:
* Underline highlighted ranges in the source text
* Make the boxes and line numbers cyan so they feel separate from the code

Here's an example from the tests:

    === main.swift ===
    1 │ let pi = 3.14159
    2 │ #myAssert(pi == 3)
      │ ╰─ in expansion of macro 'myAssert' here
      ╭-── #myAssert ──────────────────────────────────────────────────────
      │1 │ let __a = pi
      │2 │ let __b = 3
      │3 │ if #invertedEqualityCheck(__a, __b) {
      │  │    ╰─ in expansion of macro 'invertedEqualityCheck' here
      │  ╭-── #invertedEqualityCheck ─────────────────────────────────────
      │  │1 │ !(__a == __b)
      │  │  │       ╰─ no matching operator '==' for types 'Double' and 'Int'
      │  ╰-───────────────────────────────────────────────────────────────
      │4 │   fatalError("assertion failed: pi != 3")
      │5 │ }
      ╰-───────────────────────────────────────────────────────────────────
    3 │ print("hello"
      │              ╰─ expected ')' to end function call

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

File: Sources/SwiftDiagnostics/GroupedDiagnostics.swift [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      let extraLengthNeeded = targetLineLength - padding.count - sourceFile.displayName.count - 5
+      let boxSuffix: String
+      if extraLengthNeeded > 0 {
+        boxSuffix = String(repeating: "─", count: extraLengthNeeded)
+      } else {
+        boxSuffix = ""
+      }
+
+      prefixString = padding + "╭-── " + sourceFile.displayName + boxSuffix + "\n"
+      suffixString = padding + "╰-──" + String(repeating: "─", count: sourceFile.displayName.count + 1) + boxSuffix + "\n"
```
Comment (ahoppen) [inline Sources/SwiftDiagnostics/GroupedDiagnostics.swift:182]: I’d use string interpolation here.

File: Tests/SwiftParserTest/translated/StringLiteralEofTests.swift [hunk 2]
```diff
@@ -105,7 +105,7 @@ final class StringLiteralEofTests: XCTestCase {
             foo
             \(<#expression#>)"""
         """##
-      // FIXME: The closing delimiter should be put on the new line
+        // FIXME: The closing delimiter should be put on the new line
```
Comment (ahoppen) [inline Tests/SwiftParserTest/translated/StringLiteralEofTests.swift:108]: I think your local build of swift-format is too old. The formatting of this line changed with https://github.com/apple/swift-format/pull/480.

File: Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        │  │1 │ !(__a == __b)
+        │  │  ∣       ╰─ no matching operator '==' for types 'Double' and 'Int'
+        │  ╰-─────────────────────────────────────────────────────────────────
+        │4 │   fatalError("assertion failed: pi != 3")
+        │5 │ }
+        ╰-────────────────────────────────────────────────────────────────────
+      3 │ print("hello"
+        ∣              ╰─ expected ')' to end function call
+
+      """
```
Comment (ahoppen) [inline Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift:202]: The debug output looks a little bit too squashed to me. What do you think about adding an empty l line before and after starting a new source file buffer and adda leading space after the `|` of the inline source buffer. I.e.

```
      === main.swift ===
       1 │ let pi = 3.14159
       2 │ #myAssert(pi == 3)
         ∣ ╰─ in expansion of macro 'myAssert' here

         ╭-── #myAssert────────────────────────────────────────────────────────
         │ 1 │ let __a = pi
         │ 2 │ let __b = 3
         │ 3 │ if #invertedEqualityCheck(__a, __b) {
         │   ∣    ╰─ in expansion of macro 'invertedEqualityCheck' here
         │ 
         │   ╭-── #invertedEqualityCheck────────────────────────────────────────
         │   │ 1 │ !(__a == __b)
         │   │   ∣       ╰─ no matching operator '==' for types 'Double' and 'Int'
         │   ╰-─────────────────────────────────────────────────────────────────
         │ 
         │ 4 │   fatalError("assertion failed: pi != 3")
         │ 5 │ }
         ╰-────────────────────────────────────────────────────────────────────

       3 │ print("hello"
         ∣              ╰─ expected ')' to end function call

```

File: Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+final class GroupedDiagnosticsFormatterTests: XCTestCase {
+  func testGroupingForMacroExpansion() {
+    var group = GroupedDiagnostics()
+
+    // Main source file.
+    let (mainSourceID, mainSourceMarkers) = group.addTestFile(
+      """
+      let pi = 3.14159
+      0️⃣#myAssert(pi == 3)
```
Comment (ahoppen) [inline Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift:84]: Not really important but in all the other tests we start counting diagnostics with 1️⃣ not 0️⃣.

**Review by ahoppen — COMMENTED** [review]
This looks nice! I’ve got a few minor comments inline.

File: Sources/SwiftDiagnostics/GroupedDiagnostics.swift [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    /// The set of diagnostics that are emitted within this source file.
+    var diagnostics: [Diagnostic] = []
+  }
+
+  /// The set of source files for which diagnostics will be emitted.
+  var sourceFiles: [SourceFile] = []
+
+  /// Mapping from the root source file syntax nodes to the corresponding
+  /// source file IDs.
+  var rootIndexes: [SourceFileSyntax: SourceFileID] = [:]
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply Sources/SwiftDiagnostics/GroupedDiagnostics.swift:50]: Maybe? It felt more fragile than if we made this explicit.

File: Tests/SwiftParserTest/translated/StringLiteralEofTests.swift (hunk 2 above)
Comment (DougGregor) [reply Tests/SwiftParserTest/translated/StringLiteralEofTests.swift]: Ah, that makes sense.

File: Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift (hunk 3 above)
Comment (DougGregor) [reply Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift:202]: This is an easy change we can do at any point, but... this looks way too spaced out for me.  In the colorized version, I've made the boxes and line numbers cyan, which helps distinguish things more.

**Comment by DougGregor** [comment]
@swift-ci please test

File: Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift (hunk 3 above)
Comment (ahoppen) [reply Tests/SwiftDiagnosticsTest/GroupDiagnosticsFormatterTests.swift:202]: Ah, I didn’t look at the colorized output yet. If you’ve got a preference, let’s keep it as-is for now.

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift [hunk 6]
```diff
@@ -41,9 +90,108 @@ public struct DiagnosticsFormatter {
...
+
+      return startColumn..<endColumn
+    }.sorted { (lhs, rhs) in
+      lhs.lowerBound < rhs.lowerBound
+    }.mergingOverlappingRanges()
+
+    // Map the column ranges into index ranges within the source string itself.
+    let sourceString = annotatedLine.sourceString
+    let highlightIndexRanges: [Range<String.Index>] = highlightRanges.map { highlightRange in
+      let startIndex = sourceString.index(sourceString.startIndex, offsetBy: highlightRange.lowerBound - 1)
```
Comment (ahoppen) [inline Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:156]: I’m not 100% sure but I don’t think this is Unicode correct. The columns are about UTF-8 byte offsets whereas `String.Index` is about Unicode characters/grapheme clusters/something along those lines. IIRC there’s a way to convert between the two but I don’t remember what it is. Adding a few tests with emojis would certainly be good.

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift [hunk 7]
```diff
@@ -195,4 +406,14 @@ struct ANSIAnnotation {
...
   }
+
+  /// Annotation used for the outline and line numbers of a buffer.
+  static var bufferOutline: ANSIAnnotation {
+    ANSIAnnotation(color: .cyan, trait: .normal)
+  }
+
+  /// Annotation used for highlighting source text.
+  static var sourceHighlight: ANSIAnnotation {
+    ANSIAnnotation(color: .white, trait: .underline)
```
Comment (ahoppen) [inline Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:417]: Will this also print the text as white in a terminal with white background and thus make it invisible? Shouldn’t it be

```suggestion
    ANSIAnnotation(color: .normal, trait: .underline)
```

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift (hunk 6 above)
Comment (DougGregor) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:156]: Oh, I think you're right. I should be computing indices via the UTF-8 view of the source string.

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift (hunk 7 above)
Comment (DougGregor) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:417]: Yeah, I wanted to do that, but if I. use `.normal` here, Terminal doesn't show the underline.

Comment (ahoppen) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:417]: Hmm, but having invisible test in a white terminal seems like a really bad problem to me (I’m one of those weirdos who doesn’t use a dark terminal 😆).

File: Sources/SwiftDiagnostics/GroupedDiagnostics.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (DougGregor) [reply Sources/SwiftDiagnostics/GroupedDiagnostics.swift:50]: The mapping is only used in cases where the client isn't supplying a buffer ID, so you never need to depend on it. We could either drop the mapping internally (and have clients always supply the buffer ID when adding a diagnostic) or drop the buffer ID (and rely on the mapping everywhere), but this does feel like a weird middle ground.

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift (hunk 6 above)
Comment (DougGregor) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:156]: Yes, I got this fixed up in https://github.com/apple/swift-syntax/pull/1365

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift (hunk 7 above)
Comment (DougGregor) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:417]: Addressed this in https://github.com/apple/swift-syntax/pull/1365

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift [hunk 8]
```diff
@@ -41,9 +90,108 @@ public struct DiagnosticsFormatter {
...
+  ///
+  /// - Parameters:
+  ///   - suffixTexts: suffix text to be printed at the given absolute
+  ///                  locations within the source file.
+  func annotatedSource<SyntaxType: SyntaxProtocol>(
+    fileName: String?,
+    tree: SyntaxType,
+    diags: [Diagnostic],
+    indentString: String,
+    suffixTexts: [(AbsolutePosition, String)],
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:191]: This is a little nicer, so I went ahead and did it in https://github.com/apple/swift-syntax/pull/1365

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift (hunk 6 above)
Comment (ahoppen) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:156]: Thanks!

**Pull request merged**